src/api/app/controllers/webui/users/notifications_controller.rb

```diff
@@ -24,13 +26,22 @@ def update
     @count = @notifications.where(id: @notification_ids, delivered: !deliver).update_all(delivered: deliver)
     # rubocop:enable Rails/SkipsModelValidations
 
+    # update the count after the update
+    if deliver
+      @counted_notifications['unread'] -= @count
```

Comment:
This will break if either side isn't a number...

Reply:
pushed a commit to fix this

Reply:
Why do you think `@counted_notifications['unread']` could be anything different than a number?

Anyway, your solution was in my initial approach, but it does not work: if we do the counting after the update like you do now, `@notifications` is already the filtered and paginated subset, so it does not work in this straight forward way. Either we manually adjust the numbers, or we recompute the counting on the entire set, which might slow down the performance again.

Maybe we can do something in a mid-way: count again but only the unread ones (which is expected to be a fast enough query) and deduct the `read` amount by the following:
```
@counted_notifications['unread'] = User.session.unread_notifications
@counted_notifications['read'] = @counted_notifications['all'] - @counted_notifications['unread']
```

Reply:
https://github.com/openSUSE/open-build-service/pull/16257/commits/9fbac3de3ef69a00c49a2f3943dfdf8e9775885c